step-templates/upgrade_to_latest_tentacle.json

```diff
@@ -0,0 +1,104 @@
             = [bool]::Parse($OctopusParameters['Wait'])\n\n# Remove white space and blank lines.\nif ($EnvironmentNames) {\n    $EnvironmentNames = $EnvironmentNames.Split(\"`n\").Trim().Where({$_})\n}\nif ($RoleNames) {\n    $RoleNames = $RoleNames.Split(\"`n\").Trim().Where({$_})\n}\nif ($MachineNames) {\n    $MachineNames = $MachineNames.Split(\"`n\").Trim().Where({$_})\n}\n\n[Net.ServicePointManager]::SecurityProtocol = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12\n$ErrorActionPreference = \"Stop\"\n\n$header = @{ \"X-Octopus-ApiKey\" = $ApiKey }\n\nif ($SpaceName) {\n    $space = (Invoke-RestMethod -Method Get -Uri \"$OctopusUrl/api/spaces/all\" -Headers $header) | Where-Object { $_.Name -eq $SpaceName }\n    $baseUri = \"$OctopusUrl/api/$($space.Id)\"\n} else {\n    $baseUri = \"$OctopusUrl/api\"\n}\n\n# Start with no machines\n$allMachines = @()\n\n# Add machines for each requested environment.\nforeach ($environmentName in $EnvironmentNames) {\n    $environment = (Invoke-RestMethod -Method Get -Uri \"$baseUri/environments/all\" -Headers $header) | Where-Object { $_.Name -eq $environmentName }\n    $allMachines += (Invoke-RestMethod -Method Get -Uri \"$baseUri/environments/$($environment.Id)/machines?take=$([int32]::MaxValue)\" -Headers $header).Items\n}\n\n# If roles are specifed, include only machines in the specicied roles. Otherwise don't filter.\nif ($RoleNames) {\n    $roleFilteredMachines = @()\n    foreach ($roleName in $RoleNames) {\n        $roleFilteredMachines += $allMachines | Where-Object { $_.Roles -contains $roleName }\n    }\n} else {\n    $roleFilteredMachines += $allMachines\n}\n\n# Add each specific machine requested.\n$roleFilteredMachines += (Invoke-RestMethod -Method Get -Uri \"$baseUri/machines/all\" -Headers $header) | Where-Object { $_.Name -in $MachineNames }\n\n# Create array of unique IDs to target.\n$uniqueIDs = [array]($roleFilteredMachines.Id | Sort-Object -Unique)\n\nif (-not $uniqueIDs) {\n    Write-Highlight \"No machines were targeted. Exiting...\"\n    exit\n}\n\n# Build json payload, targeting unique machine IDs.\n$jsonPayload = @{\n    Arguments = @{\n        MachineIds = $uniqueIDs\n    }\n    Description = \"Upgrade Tentacle version.\"\n    Name = \"Upgrade\"\n}\n\nif ($WhatIf) {\n    Write-Host \"Upgrading tentacles on:\"\n    $roleFilteredMachines.Name | Sort-Object -Unique\n} else {\n    $task = Invoke-RestMethod -Method Post -Uri \"$baseUri/tasks\" -Headers $header -Body ($jsonPayload | ConvertTo-Json -Depth 10)\n    Write-Highlight \"$($task.Id) started. Progress can be monitored [here]($OctopusUrl$($task.Links.Web)?activeTab=taskLog)\"\n    \n    if ($Wait) {\n        do {\n        \t# Output the current state of the task every five seconds.\n            $task = Invoke-RestMethod -Method Get -Uri \"$baseUri/tasks/$($task.Id)\" -Headers $header\n            $task\n            Start-Sleep -Seconds 5\n        } while ($task.IsCompleted -eq $false)\n    }\n}"
```

Comment:
Likewise a check here to make sure `$space` is not null because the space name may not have been valid.

```
if ($SpaceName) {
    $space = (Invoke-RestMethod -Method Get -Uri "$OctopusUrl/api/spaces/all" -Headers $header) | Where-Object { $_.Name -eq $SpaceName }
    $baseUri = "$OctopusUrl/api/$($space.Id)"
} else {
    $baseUri = "$OctopusUrl/api"
}
```